### PR TITLE
feat: [IDP-1322]: IDP Licensing - License utilisation reporting

### PR DESCRIPTION
**Pull request by sathish-soundarapandian**
## Describe your changes
IDP Licensing - License utilisation reporting

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/54631)

### CODE REVIEW

File: idp-service/chart/values.yaml [hunk 1]
```diff
@@ -249,4 +249,9 @@ proxyAllowListConfigServices:
       - /token/validate
 name: idp-service
 
-internalAccounts: kmpySmUISimoRrJL6NL73w
\ No newline at end of file
+internalAccounts: kmpySmUISimoRrJL6NL73w
```
Comment (arya-harness) [inline idp-service/chart/values.yaml:252]: What is the use case of internalAccount field?

Comment (sathish-soundarapandian) [reply idp-service/chart/values.yaml:252]: @arya-harness We have an use-case in IDP module licensing usage where we need to track user count for internal accounts irrespective of the harness support user group tag. And this harness support user group tag will be used to exclude user when that user is accessing customer accounts.

Please refer to this thread for further information - https://harness.slack.com/archives/CSL96M7UG/p1696999990538369

Comment (arya-harness) [reply idp-service/chart/values.yaml:252]: @kapilgarg1996  can you review this ? 
 
 ---
 i am wondering how things would be different if it were deployed on SMP

Comment (sathish-soundarapandian) [reply idp-service/chart/values.yaml:252]: @arya-harness Just a note that these "internalAccounts" are already done. Just that it is showing in the diff as I added new lines below that.
If you can explain the complete concern over this, I can pick it up in a separate ticket after investigation.

**Comment by sathish-soundarapandian** [comment]
trigger uts

**Comment by sathish-soundarapandian** [comment]
trigger uts

**Review by pranay-harness — APPROVED** [review]

**Comment by sathish-soundarapandian** [comment]
trigger ut2, ut3, ut7

**Review by arya-harness — APPROVED** [review]

**Comment by sathish-soundarapandian** [comment]
trigger review

**Review by kapil-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54631)* status: 15 file(s) to review. 1 change requests. 1. unresolved discussions. 

___
*[`idp-service/src/main/java/io/harness/idp/license/usage/jobs/IDPTelemetryRecordsJob.java` line 39 at r2](https://reviewable.io/reviews/harness/harness-core/54631#-NhhcKNM12uYvC1T00Lp:-NhhcKNM12uYvC1T00Lq:b-acb9su) ([raw file](https://github.com/harness/harness-core/blob/ab23014163d90cc7c277dcac2d2bcb6ef665d990/idp-service/src/main/java/io/harness/idp/license/usage/jobs/IDPTelemetryRecordsJob.java#L39)):*
> ```Java
>       log.info("IDPTelemetryRecordsJob scheduler started");
>     } catch (Exception e) {
>       log.error("Exception while creating IDPTelemetryRecordsJob. Error = {}", e.getMessage(), e);
> ```

this job will run once per day then why not throw the exception? 
___
*[`idp-service/src/main/java/io/harness/idp/license/usage/reporting/IDPTelemetryPublisher.java` line 53 at r2](https://reviewable.io/reviews/harness/harness-core/54631#-Nhhd_ca5YaZxVBMcyqZ:-Nhhd_ca5YaZxVBMcyq_:bg8cm0f) ([raw file](https://github.com/harness/harness-core/blob/ab23014163d90cc7c277dcac2d2bcb6ef665d990/idp-service/src/main/java/io/harness/idp/license/usage/reporting/IDPTelemetryPublisher.java#L53)):*
> ```Java
>       accountIdList.forEach(this::sendEvent);
>     } catch (Exception e) {
>       log.error("IDPTelemetryPublisher recordTelemetry execute failed. Error = {}", e.getMessage(), e);
> ```

add the accountId for better logging 
___
*[`idp-service/src/main/java/io/harness/idp/license/usage/reporting/IDPTelemetryPublisher.java` line 70 at r2](https://reviewable.io/reviews/harness/harness-core/54631#-NhheChXC8jv7NYCpBh-:-NhheChXC8jv7NYCpBh0:bq3kysi) ([raw file](https://github.com/harness/harness-core/blob/ab23014163d90cc7c277dcac2d2bcb6ef665d990/idp-service/src/main/java/io/harness/idp/license/usage/reporting/IDPTelemetryPublisher.java#L70)):*
> ```Java
>         IDPLicenseUsageDTO idpLicenseUsageDTO =
>             idpLicenseUsage.getLicenseUsage(accountId, ModuleType.IDP, new Date().getTime(), null);
>         map.put(COUNT_ACTIVE_DEVELOPERS, idpLicenseUsageDTO.getActiveDevelopers().getCount());
> ```

make sure .getActiveDevelopers() never returns NULL
___
*[`idp-service/src/main/java/io/harness/idp/license/usage/reporting/IDPTelemetryPublisher.java` line 76 at r2](https://reviewable.io/reviews/harness/harness-core/54631#-NhhdCMA9CbWxE6a2uHD:-NhhdCMBDUrhizwm1hMY:b-tmh8rx) ([raw file](https://github.com/harness/harness-core/blob/ab23014163d90cc7c277dcac2d2bcb6ef665d990/idp-service/src/main/java/io/harness/idp/license/usage/reporting/IDPTelemetryPublisher.java#L76)):*
> ```Java
>       }
>     } catch (Exception e) {
>       log.error("Failed to send telemetry event in IDPTelemetryPublisher for account {} Error = {}", accountId,
> ```

we can throw the exception here as we are already catching it in recordTelemetry() and remove the logger from here
___
*[`idp-service/src/main/java/io/harness/idp/license/usage/repositories/IDPTelemetryStatusRepositoryCustomImpl.java` line 41 at r2](https://reviewable.io/reviews/harness/harness-core/54631#-Nhhhw2lCGUlBR07Yjvw:-Nhhhw2lCGUlBR07Yjvx:b259huy) ([raw file](https://github.com/harness/harness-core/blob/ab23014163d90cc7c277dcac2d2bcb6ef665d990/idp-service/src/main/java/io/harness/idp/license/usage/repositories/IDPTelemetryStatusRepositoryCustomImpl.java#L41)):*
> ```Java
>     IDPTelemetrySentStatus result;
>     try {
>       // Atomic lock acquiring attempt
> ```

can you pls explain how are we obtaining a lock here? as I don't see any persistentLock in place

**Review by kapil-harness — APPROVED** [review]
> Reviewed 11 of 15 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54631)* status: 15 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Comment by sathish-soundarapandian** [comment]
trigger uts

**Review by arya-harness — DISMISSED** [review]

**Comment by sathish-soundarapandian** [comment]
trigger review

**Comment by sathish-soundarapandian** [comment]
trigger uts

**Comment by sathish-soundarapandian** [comment]
trigger ut0, ut1, ut3, ut6, ut7

**Review by vikyathharekal — APPROVED** [review]
> Reviewed 10 of 15 files at r1, 1 of 1 files at r2, 2 of 2 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54631)* status: 2 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`idp-service/src/main/java/io/harness/idp/license/usage/entities/IDPTelemetrySentStatus.java` line 55 at r6](https://reviewable.io/reviews/harness/harness-core/54631#-NhvO38J3urcmhnONAh3:-NhvO38J3urcmhnONAh4:b4qi3ep) ([raw file](https://github.com/harness/harness-core/blob/49663b7145c0161c334ee274c791066bf9c4e277/idp-service/src/main/java/io/harness/idp/license/usage/entities/IDPTelemetrySentStatus.java#L55)):*
> ```Java
>     return ImmutableList.<MongoIndex>builder()
>         .add(CompoundMongoIndex.builder()
>                  .name("no_dup")
> ```

nit: unique_account_id



_Code quote:_
```Java
"no_dup"
```

**Review by pranay-harness — APPROVED** [review]

**Comment by sathish-soundarapandian** [comment]
trigger uts

**Review by arya-harness — APPROVED** [review]

**Comment by sathish-soundarapandian** [comment]
trigger uts

**Comment by sathish-soundarapandian** [comment]
trigger uts

**Comment by sandeepa-kv-jois** [comment]
Unrelated test failures, going ahead with FM

**Pull request merged**